kmip/pie/client.py

```diff
@@ -0,0 +1,389 @@
+    def __init__(self,
+                 hostname=None,
+                 port=None,
+                 cert=None,
+                 key=None,
+                 ca=None,
+                 ssl_version=None,
+                 username=None,
+                 password=None,
+                 config='client'):
+        """
+        Construct a ProxyKmipClient.
+
+        Args:
+            hostname (string): The host or IP address of a KMIP appliance.
+                Optional, defaults to None.
+            port (int): The port number used to establish a connection to a
+                KMIP appliance. Usually 5696 for KMIP applications. Optional,
+                defaults to None.
+            cert (string): The path to the client's certificate. Optional,
+                defaults to None.
+            key (string): The path to the key for the client's certificate.
+                Optional, defaults to None.
+            ca (string): The path to the CA certificate used to verify the
+                server's certificate. Optional, defaults to None.
+            ssl_version (string): The name of the ssl version to use for the
+                connection. Example: 'PROTOCOL_SSLv23'. Optional, defaults to
+                None.
+            username (string): The username of the KMIP appliance account to
+                use for operations. Optional, defaults to None.
+            password (string): The password of the KMIP appliance account to
+                use for operations. Optional, defaults to None.
+            config (string): The name of a section in the PyKMIP configuration
+                file. Use to load a specific set of configuration settings from
+                the configuration file, instead of specifying them manually.
+                Optional, defaults to the default client section, 'client'.
+        """
+        self.logger = logging.getLogger()
+
+        self.attribute_factory = attributes.AttributeFactory()
+        self.object_factory = factory.ObjectFactory()
+
+        # TODO (peter-hamilton) Consider adding validation checks for inputs.
+        self.proxy = KMIPProxy(
+            host=hostname,
+            port=port,
+            certfile=cert,
+            keyfile=key,
+            ca_certs=ca,
+            ssl_version=ssl_version,
+            username=username,
+            password=password,
+            config=config)
+
+        self._is_open = False
+        self._lock = multiprocessing.RLock()
+
+    def _synchronize(f):
+        """
+        A synchronization decorator for creating thread-safe client methods.
+
+        Args:
+            f (function): The function to be wrapped in synchronization checks.
+
+        Returns:
+            function: The original function protected by a synchronization
+                lock.
+        """
+        def wrapper(self, *args, **kwargs):
+            with self._lock:
+                return f(self, *args, **kwargs)
+
+        return wrapper
+
+    @_synchronize
```

Comment:
Just so everyone's on the same page, I left the synchronization code in (for now) because I still wanted feedback on my questions (way above) regarding using a separate `utils` module or `oslo.concurrency`, as @joel-coffman suggested. If we want to defer this conversation, I'm happy to pull the synchronization out.